Add additional sensors for Arlo Baby camera

lukiffer opened the pull request:
## Description:
Adds additional temperature, humidity, and air quality sensors for Arlo Baby cameras implemented in `pyarlo==0.1.8`. 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5583

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: arlo
    monitored_conditions:
      - temperature
      - humidity
      - air_quality
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

lukiffer committed:
Add additional sensors for Arlo Baby camera

lukiffer committed:
Fix linter errors

MartinHjelmare commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -124,13 +153,31 @@ def update(self):
             try:
                 self._state = self._data.battery_level
-            except TypeError:
+            except (AttributeError, TypeError):
```
Why would there be an `AttributeError`?

tchellomello replied:
Yeah I don't think the attribute error should be necessary

MartinHjelmare reviewed (commented):
@tchellomello could you have a look and review?

tchellomello commented:
It looks good to me. I don't have a baby monitor but code looks good!

lukiffer commented:
This PR may have been slightly premature. I’ve since discovered an issue with stats not updating. I’m working on resolving this before removing the WIP designation. If it would be preferable to close this PR and reopen one once this issue is resolved, feel free to close this.

tchellomello reviewed (changes requested):
@lukiffer  could you bump the Pyarlo module to  `pyarlo==0.1.9` . 

https://pypi.org/project/pyarlo/0.1.9/

Thanks!

lukiffer committed:
Fix linter error

lukiffer committed:
Add tests for Arlo sensors

lukiffer committed:
Merge branch 'dev' of github.com:home-assistant/home-assistant into arlo-baby

lukiffer committed:
Fix linter errors

lukiffer committed:
Bump pyarlo dependency to 0.1.9

lukiffer committed:
Remove unnecessary AttributeError except

lukiffer committed:
Fix module reference error in py35

lukiffer committed:
Fix test

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,185 @@
+
+    def test_sensor_name(self):
+        """Test the name property."""
+        sensor = TestArloSensor._get_mock_sensor()
```
The method is weirdly named since it doesn't seem to return a mock sensor but an actual instance of an arlo sensor.

Access methods using `self.`.

lukiffer replied:
Will correct the method name and change to `self` access.

Also, just for reference as I'm coming back to python several years away, is it not optimal to use the `@staticmethod` decorator and treat these as "static" as opposed to members? Or is this just convention? Or is this an artifact of having this in a class as opposed to just standalone functions? Thanks for the help.

MartinHjelmare replied:
Static methods can still be accessed on the instance, ie using `self.` inside instance methods. It's only inside the static method that the instance isn't available.

Yeah, the static methods are part of what could be cleaned up if we would move to standalone pytest tests.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,185 @@
+        sensor = TestArloSensor._get_mock_sensor()
+        self.assertEqual(sensor.name, 'Last')
+
+    @asyncio.coroutine
```
Remove this.

lukiffer replied:
Again, having just come back to python (after several years and from python2), async methods in python3 have come quite a long way, and I'm still getting accustomed to the new stuff:

When I omit this, it complains that the function is never awaited. In tests for other HA components, I noticed that standalone functions omit this, but functions declared in classes have it. Is this correlation true or just coincidence? If it does in fact need to be removed, what's the appropriate way to address the no-await issue?

MartinHjelmare replied:
We're already declaring the method a coroutine by saying `async def`. This is the new async syntax in Python 3.5. So we shouldn't add the coroutine decorator on top of that. Class based unittest tests don't support coroutines as test methods. Here we have to move to standalone pytest tests.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,185 @@
+        """Test dispatcher called when added."""
+        sensor = TestArloSensor._get_mock_sensor()
+        await sensor.async_added_to_hass()
+        dispatcher.async_dispatcher_connect.assert_called_once()
```
We prefer not to use `assert_called_once` but instead use `assert len(mock.calls) == 1`. A typo on the mock method would pass silently.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,185 @@
+
+    def test_sensor_state_default(self):
+        """Test the state property."""
+        sensor = TestArloSensor._get_mock_sensor()
```
Use `self._get_mock_sensor()`.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,185 @@
+    sensors = None
+
+    def _add_devices(self, sensors, boolean):
+        if boolean:
```
Why would we only add the sensors when passing `True` as the second argument? That's not how the actual code works.

lukiffer replied:
Not actually sure what the second param does -- just noticed that it's hardcoded `True` in the sensor code. But this does make the test rather brittle and potentially (probably) incorrect. I've removed the condition for now and will take another pass after some discovery.

MartinHjelmare replied:
Passing `True` to `add_devices` calls `entity.update` during entity addition.

MartinHjelmare reviewed (commented):
I think the tests would be cleaner if they would be converted to pytest standalone  test functions and using pytest fixtures. But it's not a requirement from my side.

lukiffer commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -133,6 +164,24 @@ def update(self):
+        elif self._sensor_type == 'temperature':
+            try:
+                self._state = self._data.ambient_temperature
+            except (AttributeError, TypeError):
```
I had these in there because the library was in some cases bubbling up an `AttributeError` but in further testing, it appears to be the result of an issue that was fixed in `pyarlo==0.1.9` where the method to get the ambient sensor data was not being called on update. I've removed them and will push.

lukiffer commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -141,10 +190,16 @@ def device_state_attributes(self):
-           self._sensor_type == 'captured_today' or \
-           self._sensor_type == 'battery_level' or \
-           self._sensor_type == 'signal_strength':
+        known_sensor_types = [
```
This was a lazy workaround for address a maximum conditions in an if statement on my part. There's a dictionary with keys of known types already at the module level, I'll use that.

lukiffer committed:
Address PR review concerns

lukiffer committed:
Convert to standalone pytest methods

lukiffer committed:
Fix linter errors

lukiffer committed:
Fix linter errors

lukiffer committed:
Fix linter errors

lukiffer committed:
Fix test

MartinHjelmare commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -141,10 +189,12 @@ def device_state_attributes(self):
-           self._sensor_type == 'signal_strength':
+        known = False
+        for sensor_type in SENSOR_TYPES:
+            if self._sensor_type == sensor_type:
```
We already validate this in the config schema. We can remove this check completely here.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,204 @@
+def test_async_added_to_hass(default_sensor):
+    """Test dispatcher called when added."""
+    yield from default_sensor.async_added_to_hass()
+    assert len(dispatcher.async_dispatcher_connect.calls) == 1
```
We should check that the calls on the mock object is of correct length.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,204 @@
+
+@patch('homeassistant.helpers.dispatcher.async_dispatcher_connect',
+       MagicMock())
+def test_async_added_to_hass(default_sensor):
```
Define this as a coroutine with `async def`.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,204 @@
+       MagicMock())
+def test_async_added_to_hass(default_sensor):
+    """Test dispatcher called when added."""
+    yield from default_sensor.async_added_to_hass()
```
Use `await` instead of `yield from`.

MartinHjelmare commented on tests/components/sensor/test_arlo.py:
```diff
@@ -0,0 +1,204 @@
+    assert default_sensor.name == 'Last'
+
+
+@patch('homeassistant.helpers.dispatcher.async_dispatcher_connect',
```
Move this to a fixture in which we yield the mock.

lukiffer committed:
Remove redundant check, fix async test

lukiffer committed:
Fix linter error

lukiffer commented:
@MartinHjelmare thanks again for the feedback. I fixed a problem where I was patching the wrong namespace -- this helped make sense of a lot of the other feedback you gave. I did however still have one thing I was stuck on -- not sure if it's a blocker though:

I removed the `@patch()` decorator from the test method intending to move it onto a fixture, but am having an issue with passing the mock by reference on the fixture. When using the `with patch() as` in the test method, the test works correctly and I've added checks for the `call_args`.

However, when I pass the sensor and the mock as a fixture, e.g.:

```
class DispatcherFixture():
    def __init__(self, sensor, mock):
        self.sensor = sensor
        self.mock = mock

@patch('homeassistant.components.sensor.arlo.async_dispatcher_connect',
        MagicMock())
@pytest.fixture()
def dispatcher_fixture(default_sensor, hass):
    # ...
    return DispatcherFixture(default_sensor, arlo.async_dispatcher_connect)
```

I get `dispatcher_fixture.mock` as an instance of `MagicMock` but the only call is `[call.__eq__(sentinel.DEFAULT)]`. Not sure what's going awry here, but again the test is functional as committed and I don't foresee the fixture being reused for another test.

As an aside, this may be getting a bit out of my league, and given that the code under test (for this specific test) isn't part of the scope of the initial feature addition, perhaps it would be appropriate to redact this test from this PR and file a separate PR with only the appropriate/pythonic implementation of this test. What are your thoughts?

MartinHjelmare commented:
It's ok to keep the test as is.

This is how you could have applied the patch within a fixture. Note the `yield`.

```py
@pytest.fixture
def mock_dispatch():
    with patch('path.to.dispatch') as _mock:
        yield _mock
```

MartinHjelmare commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -133,18 +163,31 @@ def update(self):
-           self._sensor_type == 'battery_level' or \
-           self._sensor_type == 'signal_strength':
-            attrs['model'] = self._data.model_id
+        attrs['model'] = self._data.model_id
```
Add a test or update the test where all sensor types are set up, and test that attributes return ok also for the total cameras type. I think we need to add a check here where we don't add the model attribute for that sensor type.

MartinHjelmare reviewed (commented):
It's easier to catch bugs if we set up the complete platform, while patching appropriately, and run all checks on the entity using the home assistant core, eg get the state from the state machine and check that it's ok. Ie write the tests more as integration tests than isolated unit tests.

lukiffer committed:
Added check for total_cameras sensor, added additional attribute tests

lukiffer committed:
Add missing docstring

lukiffer commented:
@MartinHjelmare I've added the additional test cases you mentioned and updated the dispatcher test to use a fixture -- thanks for giving me the last piece of the puzzle on that one (again). I also added the check for total_cameras sensor type -- good catch.

I'm pretty sure that what you mentioned about having fully-setup integration tests as opposed to unit tests is not satisfied by the latest commits. I can definitely see the benefit of this approach (especially given the total_cameras example you highlighted). I would like to potentially focus on that a bit more in a separate branch/PR given the scope of those changes. Is that acceptable or should I proceed in this branch?

MartinHjelmare reviewed (approved):
Looks good!

lukiffer commented:
@MartinHjelmare thanks again for all your help with this. @tchellomello could you also give these changes another look? Thanks!

tchellomello reviewed (approved):
Looks good @lukiffer!! Thanks for your contribution!!

Thanks @MartinHjelmare for the code review!!
